system-addon/test/functional/mochitest/browser_topsites_section.js

```diff
@@ -0,0 +1,64 @@
+"use strict";
+
+registerCleanupFunction(async () => {
+  Services.prefs.clearUserPref("browser.newtabpage.activity-stream.default.sites");
+  Services.prefs.clearUserPref("browser.newtabpage.pinned");
+  await clearHistoryAndBookmarks();
+});
+
+// Check TopSites edit modal and overlay show up.
+test_newtab({
+  async before() {
+    await clearHistoryAndBookmarks();
+  },
+  // it should be able to click the topsites edit button to reveal the edit topsites modal and overlay.
+  test: function topsites_edit() {
+    const topsitesEditBtn = content.document.querySelector(".edit-topsites-button button");
+    topsitesEditBtn.click();
+
+    let found = content.document.querySelector(".edit-topsites");
+    ok(found && !found.hidden, "Should find a visible topsites edit menu");
+
+    found = content.document.querySelector(".modal-overlay");
+    ok(found && !found.hidden, "Should find a visible overlay");
+  }
+});
+
+// Test pin/unpin context menu options.
+test_newtab({
+  async before() {
+    // The pref for TopSites is empty by default.
+    Services.prefs.setStringPref("browser.newtabpage.activity-stream.default.sites", "https://www.youtube.com/,https://www.facebook.com/,https://www.amazon.com/,https://www.reddit.com/,https://www.wikipedia.org/,https://twitter.com/");
+    await clearHistoryAndBookmarks();
+  },
+  // it should pin the website when we click the first option of the topsite context menu.
+  test: async function topsites_pin_unpin() {
+    // There are only topsites on the page, the selector with find the first topsite menu button.
+    const topsiteContextBtn = content.document.querySelector(".context-menu-button");
+    topsiteContextBtn.click();
+
+    const contextMenu = content.document.querySelector(".context-menu");
+    ok(contextMenu && !contextMenu.hidden, "Should find a visible topsite context menu");
+
+    const pinUnpinTopsiteBtn = contextMenu.querySelector(".context-menu-item a");
+    // Pin the topsite.
+    pinUnpinTopsiteBtn.click();
+
+    // Need to wait for pin action.
+    await ContentTaskUtils.waitForCondition(() => content.document.querySelector(".icon-pin-small"),
+      "No highlights cards found.");
+
+    let pinnedIcon = content.document.querySelectorAll(".icon-pin-small").length;
+    is(pinnedIcon, 1, "should find 1 pinned topsite");
+
+    // Unpin the topsite.
+    pinUnpinTopsiteBtn.click();
+
+    // Need to wait for unpin action.
+    await ContentTaskUtils.waitForCondition(() => content.document.querySelectorAll(".icon-pin-small").length === 0,
+      "No highlights cards found.");
+
+    pinnedIcon = content.document.querySelectorAll(".icon-pin-small").length;
+    is(pinnedIcon, 0, "should find no pinned topsites");
```

Comment:
This check seems to be implied by the previous `wait`. At least the earlier check for 1 `length` is to make sure we didn't get multiple pinned. Also, the condition should be able to just be `!document.querySelector(".icon-pin-small")` with appropriate failure message.